aspnetcore/fundamentals/dependency-injection.md

```diff
@@ -150,36 +150,78 @@ public void ConfigureServices(IServiceCollection services)
 
 > [!WARNING]
 > It's dangerous to resolve a scoped service from a singleton. It may cause the service to have incorrect state when processing subsequent requests.
 
+## Service registration methods
+
+Each service registration extension method offers overloads that are useful in specific scenarios.
+
+| Method | Automatic<br>object<br>disposal | Mocks for<br>testing | Pass args |
+| ------ | :-----------------------------: | :------------------: | :-------: |
+| `Add{LIFETIME}<{INTERFACE}, {TYPE}>()`<br>Example:<br>`services.AddScoped<IMyDep, MyDep>();` | Yes | Yes | No |
+| `Add{LIFETIME}<{INTERFACE}>(sp => new {TYPE})`<br>Examples:<br>`services.AddScoped<IMyDep>(sp => new MyDep());`<br>`services.AddScoped<IMyDep>(sp => new MyDep("A string!"));` | Yes | Yes | Yes |
+| `Add{LIFETIME}<{TYPE}>()`<br>Example:<br>`services.AddScoped<MyDep>();` | Yes | No | No |
+| `Add{LIFETIME}<{INTERFACE}>(new {TYPE})`<br>Examples:<br>`services.AddScoped<IMyDep>(new MyDep());`<br>`services.AddScoped<IMyDep>(new MyDep("A string!"));` | No | Yes | Yes |
+| `Add{LIFETIME}(new {TYPE})`<br>Examples:<br>`services.AddScoped(new MyDep());`<br>`services.AddScoped(new MyDep("A string!"));` | No | No | Yes |
+
+For more information on type disposal, see the [Disposal of services](#disposal-of-services) section. For more information on mocking types for testing, see <xref:test/integration-tests#inject-mock-services>.
+
+`TryAdd{LIFETIME}` methods only register the service if there isn't already an implementation registered for the service.
+
+In the following example, the first line registers `MyDependency` for `IMyDependency`. The second line no-ops because `IMyDependency` already has a registered type:
+
+```csharp
+services.AddSingleton<IMyDependency, MyDependency>();
+// The following line no-ops:
+services.TryAddSingleton<IMyDependency, DifferentDependency>();
+```
+
+For more information, see:
+
+* <xref:Microsoft.Extensions.DependencyInjection.Extensions.ServiceCollectionDescriptorExtensions.TryAdd*>
+* <xref:Microsoft.Extensions.DependencyInjection.Extensions.ServiceCollectionDescriptorExtensions.TryAddTransient*>
+* <xref:Microsoft.Extensions.DependencyInjection.Extensions.ServiceCollectionDescriptorExtensions.TryAddScoped*>
+* <xref:Microsoft.Extensions.DependencyInjection.Extensions.ServiceCollectionDescriptorExtensions.TryAddSingleton*>
+
+[TryAddEnumerable(ServiceDescriptor)](xref:Microsoft.Extensions.DependencyInjection.Extensions.ServiceCollectionDescriptorExtensions.TryAddEnumerable*) methods only register the service if there isn't already a registered instance of a specific implementing type. Multiple services are resolved via `IEnumerable<{INTERFACE}>`; but when registering services, the the developer only wants to add an instance if one of the same type hasn't already been added. Generally, this method is used by library authors to avoid registering two copies of an instance in the container.
```

Comment:
```suggestion
[TryAddEnumerable(ServiceDescriptor)](xref:Microsoft.Extensions.DependencyInjection.Extensions.ServiceCollectionDescriptorExtensions.TryAddEnumerable*) methods only register the service if there isn't already an implementation **of the same type**. Multiple services are resolved via `IEnumerable<{SERVICE}>`; but when registering services, the the developer only wants to add an instance if one of the same type hasn't already been added. Generally, this method is used by library authors to avoid registering two copies of an instance in the container.
```

The thing I'm trying to clarify here is that **both** the service type and implementation type have to match for `TryAddEnumerable` to no-op. I may be fumbling for words a bit 😄. Also, maybe too esoteric to worry about here unless confused customers come knocking :)

For example:

```csharp
public interface IFoo {}
public interface IBar {}

public class FooBar: IFoo, IBar {}

// Registers an instance of FooBar
services.TryAddEnumerable(ServiceDescriptor.Singleton<IFoo, FooBar>());
// Registers a SECOND instance of FooBar with it's own lifetime
services.TryAddEnumerable(ServiceDescriptor.Singleton<IBar, FooBar>());
// No-ops because an (IFoo, FooBar) pair already exists.
services.TryAddEnumerable(ServiceDescriptor.Singleton<IFoo, FooBar>());
```

Reply:
Even if we don't take that full change, we should change to `IEnumerable<{SERVICE}>` as per my comments above about prefering "service" over "implementation"